File: src-tauri/src/commands/verification.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationData {
    pub attestation_number: String,
    pub date_verification: String,
    pub immatriculation: String,
    pub date_effet: String,
    pub date_echeance: String,
    pub marque: String,
    pub modele: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationResponse {
    pub operation_status: String,
    pub operation_message: String,
    pub data: Option<VerificationData>,
}

#[tauri::command]
pub async fn verify_contract(immatriculation: String) -> Result<VerificationResponse, String> {
    let normalized = immatriculation
        .chars()
        .filter(|c| c.is_ascii_alphanumeric())
        .collect::<String>()
        .to_uppercase();

    if normalized.is_empty() {
        return Err("Immatriculation invalide.".to_string());
    }

    let url = format!(
        "https://apiaas.diotali.com/applicationtiers/verify/{}",
        normalized
    );

    let client = reqwest::Client::new();
    let response = client
        .get(url)
        .send()
        .await
        .map_err(|e| format!("Erreur réseau: {e}"))?;

    if !response.status().is_success() {
        return Err(format!("Erreur API: HTTP {}", response.status()));
    }

    response
        .json::<VerificationResponse>()
        .await
        .map_err(|e| format!("Réponse API invalide: {e}"))
}

File: server/src/routes/verification.ts
/**
 * Proxy de verification d'un contrat auprès de l'API AAS.
 * Le handler est partagé entre la route publique et la route authentifiée
 * historique afin de conserver exactement les mêmes validations.
 */

import type { Context } from "hono";
import { z } from "zod";

const verifyParamSchema = z.string().min(1).max(20);

const verificationDataSchema = z.object({
  attestationNumber: z.string(),
  dateVerification: z.string(),
  immatriculation: z.string(),
  dateEffet: z.string(),
  dateEcheance: z.string(),
  marque: z.string(),
  modele: z.string(),
});

const verificationResponseSchema = z.object({
  operationStatus: z.enum(["SUCCESS", "ERROR"]),
  operationMessage: z.string(),
  data: verificationDataSchema.nullable(),
});

const VERIFY_TIMEOUT_MS = 15_000;

export async function verificationHandler(c: Context): Promise<Response> {
  const parsed = verifyParamSchema.safeParse(c.req.param("immatriculation"));
  if (!parsed.success) return c.json({ error: "Immatriculation invalide" }, 400);

  const normalized = parsed.data.replace(/[^a-zA-Z0-9]/g, "").toUpperCase();
  if (!normalized) return c.json({ error: "Immatriculation invalide" }, 400);

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), VERIFY_TIMEOUT_MS);
  try {
    const upstream = await fetch(
      `https://apiaas.diotali.com/applicationtiers/verify/${normalized}`,
      { signal: controller.signal },
    );
    if (!upstream.ok) {
      return c.json({ error: `Erreur API : HTTP ${upstream.status}` }, 502);
    }

    const body = await upstream.json().catch(() => null);
    const validated = verificationResponseSchema.safeParse(body);
    if (!validated.success) {
      return c.json({ error: "Réponse API invalide" }, 502);
    }

    return c.json(validated.data);
  } catch (err) {
    const message = err instanceof Error ? err.message : "Erreur réseau";
    return c.json({ error: `Erreur réseau : ${message}` }, 502);
  } finally {
    clearTimeout(timeout);
  }
}
